src/iter/find_first_last/test.rs

```diff
@@ -68,3 +68,82 @@ fn same_range_last_consumers_return_correct_answer() {
     assert_eq!(reducer.reduce(left_folder.complete(), right_folder.complete()),
                Some(2));
 }
+
+// These tests requires that a folder be assigned to an iterator with more than
+// one element. We can't necessarily determine when that will happen for a given
+// input to find_first/find_last, so we test the folder directly here instead.
+#[test]
+fn find_first_folder_does_not_clobber_first_found() {
+    let best_found = AtomicUsize::new(usize::max_value());
+    let f = FindFolder {
+        find_op: &(|&_: &i32| -> bool { true }),
+        boundary: 0,
+        match_position: MatchPosition::Leftmost,
+        best_found: &best_found,
+        item: None,
+    };
+    let f = f.consume(0_i32).consume(1_i32).consume(2_i32);
+    assert!(f.full());
+    assert_eq!(f.complete(), Some(0_i32));
+}
+
+#[test]
+fn find_last_folder_yields_last_match() {
+    let best_found = AtomicUsize::new(0);
+    let f = FindFolder {
+        find_op: &(|&_: &i32| -> bool { true }),
+        boundary: 0,
+        match_position: MatchPosition::Rightmost,
+        best_found: &best_found,
+        item: None,
+    };
+    let f = f.consume(0_i32).consume(1_i32).consume(2_i32);
+    assert_eq!(f.complete(), Some(2_i32));
+}
+
+
+/// Produce a parallel iterator for 0u128..10²⁷
+fn octillion() -> impl ParallelIterator<Item = u128> {
+    (0u32..1_000_000_000)
+        .into_par_iter()
+        .with_max_len(1_000)
+        .map(|i| i as u128 * 1_000_000_000)
+        .flat_map(
+            |i| {
+                (0u32..1_000_000_000)
+                    .into_par_iter()
+                    .with_max_len(1_000)
+                    .map(move |j| i + j as u128)
+            }
+        )
+        .map(|i| i as u128 * 1_000_000_000)
+        .flat_map(
+            |i| {
+                (0u32..1_000_000_000)
+                    .into_par_iter()
+                    .with_max_len(1_000)
+                    .map(move |j| i + j as u128)
+            }
+        )
+}
+
+#[test]
+fn find_first_octillion() {
+    let x = octillion().find_first(|_| true);
+    assert_eq!(x, Some(0));
+}
+
+#[test]
+fn find_last_octillion() {
+    // FIXME: If we don't use at least two threads, then we end up walking
+    // through the entire iterator sequentially, without the benefit of any
+    // short-circuiting.  We probably don't want testing to wait that long. ;)
+    // It would be nice if `find_last` could prioritize the later splits,
+    // basically flipping the `join` args, without needing indexed `rev`.
+    // (or could we have an unindexed `rev`?)
```

Comment:
Right.  I'd like this sort of thing to be usable even in a single-threaded "pool", and walking from start to end is a poor way to find the last match.  I doubt it would come up often with absurd iterators like this, but it should be a good optimization to search roughly backwards in general.